### DISCUSSION TITLE
Local Connections only between NJSPC and REM

### DISCUSSION THREAD

**Discussion by Telemachus1977** [tagyoureit/nodejs-poolController#615] [App and O/S Config Questions] [answered]
I'd like REM and NJS-PC to continue to communicate without being connected to a router.  This would be useful if they cat chews the router cord, or if I want to stick the pi somewhere without any wifi signal, but still have it run it's automation.

Where the connections are specified in REM, I have tried using the local IP address 127.0.0.1:4200, but within seconds the IP address assigned by the router gets automatically added. 

When the wifi signal is removed, it appears that all connection between the two is lost.

Am i missing something obvious?

**Comment by rstrouse** [comment]
> Where the connections are specified in REM, I have tried using the local IP address 127.0.0.1:4200, but within seconds the IP address assigned by the router gets automatically added.

Where are you referring to?  Did you set the ip address on the RelayEquipmentManager interface to the loopback?

**Reply by Telemachus1977** [reply]
![image](https://user-images.githubusercontent.com/25725743/190921458-fada8a73-db63-4a91-b00c-e9cb74fe22e9.png)

This is what I tried.  The second one gets added automatically (I thought I remembered seeing an option somewhere to stop automatic adding?).

Should that work with no wifi connection?

**Comment by rstrouse** [comment, marked as answer]
> Where the connections are specified in REM, I have tried using the local IP address 127.0.0.1:4200, but within seconds the IP address assigned by the router gets automatically added.

Where are you referring to?  Did you set the ip address on the RelayEquipmentManager interface to the loopback?

**Reply by rstrouse** [reply]
You need to go into njsPC and set the IP address on the REM interface.

**Reply by Telemachus1977** [reply]
![Screenshot_20220918_125217_VNC Viewer](https://user-images.githubusercontent.com/25725743/190925808-1e354653-52d2-4f4e-a03e-5b8bac6b6fdf.jpg)

Like this, correct?  So with this you would expect it to work with no wifi connection, right?

**Reply by rstrouse** [reply]
Yes

**Reply by Telemachus1977** [reply]
Indeed it does.  Is this the preferred way to set it up when njsPC and REM are on the same device directly controlling equipment?

When "Find REM controllers" is selected it would be great if "local controller" popped up as an option.  Although typing in the local IP address is not too difficult.

**Reply by rstrouse** [reply]
The setup s entirely up to you.  You could put njsPC in your server rack and simply put another microprocessor at the pad for REM.  You can also put dashPanel another server.  It is a micro services architecture.

**Reply by Telemachus1977** [reply]
Very cool. 

In the not-so-distant past I remember there being an option in njsPC controller>interfaces>REM to choose whether the feed was automatically added to the REM Connections tab.  Now it just adds it no matter what.

But it adds the router assigned IP address even though I have designated a local one.

I subsequently delete it, since I already have it pointed at local in REM.

Is there a way to get it to stop adding it?

**Reply by rstrouse** [reply]
Pull njsPC and REM it will now check for a local loopback connection if you use localhost or 127.0.0.1 as the interface address.  If a loopback socket does not exist it will add it but if it does already exist it will not.

**Reply by tagyoureit** [reply]
@Telemachus1977 I think you are thinking about the ChemController setting... (I haven't pulled the new code yet so maybe @rstrouse changed this in the last 5 minutes.  😄  )

<img width="632" alt="2022-09-19_10-52-38" src="https://user-images.githubusercontent.com/1646840/191081959-441ea411-9ba6-47c7-a9c6-e6c1e9ef582c.png">

**Reply by rstrouse** [reply]
This is more about the connections list on the General tab for REM.  It attempts to set up a socket connection so REM can call back to njsPC in the feeds.  Previously, it used the ip address or host that was assigned to the njsPC server and dynamically created the connection object in REM.  While this is great, the problem comes in when you want to assign the comms to the loopback (127.0.0.1 [localhost]).   

Since dhcp assigned an address to the njsPC instance it tries to use this address for the new socket connection definition even though the loopback has been assigned to the interface.  Now the code looks at the interface definition and retains the loopback settings on the REM side.  It doesn't make sense to add another option to the interface definition since the loopback is easily detected and deterministic for the new connection.

**Reply by Telemachus1977** [reply]
@tagyoureit I think you are right, that's what I was thinking of.

@rstrouse I just played with it for several minutes, adding removing localhost address.  Problem is fixed.

Could the software show the loopback as an option when the "find" button is pressed, and add the localhost address?  Or can it only show the router assigned addresses?

**Reply by rstrouse** [reply]
It is doing an ssdp lookup and you cannot respond to a broadcast packet over the loopback.  It would take quite a bit of work to determine if the datagram came from the local adapter and it is an infection vector if njsPC responds with the loopback.  Any self respecting router should toss that in the bit bucket and suspend udp.

I'll keep that spinning around in the noggin and maybe something will fall out.  Maybe the only way to get the information is to try to ping REM on the loopback and ignore upnp.  Simply try to get `http://localhost:8080/upnp.xml` and see what comes back.

**Reply by Telemachus1977** [reply]
Well inserting local host certainly is not difficult to do manually.   

If people aren't very familiar with IP addresses or what is going on here, they will not realize that there are automation is going to shut down if they lose connection to their router.

If nothing else, maybe just a little comment message to inform them.

**Reply by Telemachus1977** [reply]
Maybe just a button to  add localhost that shows up below the list of ssdp lookup results?
